bookwyrm/models/bookwyrm_export_job.py

```diff
@@ -0,0 +1,217 @@
+    """Generate an export for a user"""
+    # user
+    exported_user = {}
+    vals = [
+        "username",
+        "name",
+        "summary",
+        "manually_approves_followers",
+        "hide_follows",
+        "show_goal",
+        "show_suggested_users",
+        "discoverable",
+        "preferred_timezone",
+        "default_post_privacy",
+    ]
+    for k in vals:
+        exported_user[k] = getattr(user, k)
+
+    if getattr(user, "avatar", False):
+        exported_user["avatar"] = f'https://{DOMAIN}{getattr(user, "avatar").url}'
+
+    # reading goals
+    reading_goals = models.AnnualGoal.objects.filter(user=user).distinct()
+    goals_list = []
+    # TODO: either error checking should be more sophisticated
+    # or maybe we don't need this try/except
+    try:
+        for goal in reading_goals:
+            goals_list.append(
+                {"goal": goal.goal, "year": goal.year, "privacy": goal.privacy}
+            )
+    except Exception:  # pylint: disable=broad-except
+        pass
+
+    try:
+        readthroughs = models.ReadThrough.objects.filter(user=user).distinct().values()
+        readthroughs = list(readthroughs)
+    except Exception:  # pylint: disable=broad-except
+        readthroughs = []
+
+    # books
+    editions, books = get_books_for_user(user)
+    final_books = []
+
+    for book in books.values():
+        edition = editions.filter(id=book["id"])
+        book["edition"] = edition.values()[0]
+        # authors
+        book["authors"] = list(edition.first().authors.all().values())
+        # readthroughs
+        book_readthroughs = (
+            models.ReadThrough.objects.filter(user=user, book=book["id"])
+            .distinct()
+            .values()
+        )
+        book["readthroughs"] = list(book_readthroughs)
+        # shelves
+        shelf_books = models.ShelfBook.objects.filter(
+            user=user, book=book["id"]
+        ).distinct()
+        shelves_from_books = models.Shelf.objects.filter(
+            shelfbook__in=shelf_books, user=user
+        )
+
+        book["shelves"] = list(shelves_from_books.values())
+        book["shelf_books"] = {}
+
+        for shelf in shelves_from_books:
+            shelf_contents = models.ShelfBook.objects.filter(
+                user=user, shelf=shelf
+            ).distinct()
+
+            book["shelf_books"][shelf.identifier] = list(shelf_contents.values())
+
+        # book lists
+        book_lists = models.List.objects.filter(
+            books__in=[book["id"]], user=user
+        ).distinct()
+        book["lists"] = list(book_lists.values())
+        book["list_items"] = {}
+        for blist in book_lists:
+            list_items = models.ListItem.objects.filter(book_list=blist).distinct()
+            book["list_items"][blist.name] = list(list_items.values())
+
+        # reviews
+        reviews = models.Review.objects.filter(user=user, book=book["id"]).distinct()
```

Comment:
Would it work to use `models.Status.objects.select_subclasses.filter(...)` to collect all the subclasses, rather than doing them separately?

Reply:
Ooh. I forgot that existed. Uh, probably? We'd need to make sure we still store them separately or at least retain knowledge of the underlying model though.

Reply:
I took a look at this but it seemed to make things more complicated? I can see that it would reduce the database queries, but since we still need to know what class everything is and get it into JSON, I'm struggling to see how to use it in a way that makes things simpler rather than more complex. Advice welcome.

Reply:
I think it was just not a helpful suggestion on my part and you can skip it